tests/install/test_builder.py

```diff
@@ -0,0 +1,208 @@
+        return _partial_partial
+    return _partial
+
+_confirm_optional_dep_count      = lambda t: lambda should_build, xs: _confirm_dep_count(t)(should_build)(xs)
+confirm_dlpack_dep_count         = _confirm_dep_count(build._DLPackRepository)(True)
+confirm_pt_dep_count             = _confirm_optional_dep_count(build._PTArchive)
+confirm_tf_dep_count             = _confirm_optional_dep_count(build._TFArchive)
+confirm_ort_dep_count            = _confirm_optional_dep_count(build._ORTArchive)
+# fmt: on
+
+
+@for_each_device
+@toggle_build_tf
+@toggle_build_pt
+@toggle_build_ort
+def test_rai_builder_will_add_dep_if_backend_requested_wo_duplicates(
+    device, build_tf, build_pt, build_ort
+):
+    rai_builder = build.RedisAIBuilder(
+        build_tf=build_tf, build_torch=build_pt, build_onnx=build_ort
+    )
+    requested_backends = rai_builder._get_deps_to_fetch_for(device)
+    assert confirm_dlpack_dep_count(requested_backends)
+    assert confirm_tf_dep_count(build_tf, requested_backends)
+    assert confirm_pt_dep_count(build_pt, requested_backends)
+    assert confirm_ort_dep_count(build_ort, requested_backends)
+
+
+@for_each_device
+@toggle_build_tf
+@toggle_build_pt
+def test_rai_builder_will_not_add_dep_if_custom_dep_path_provided(
+    device, p_test_dir, build_tf, build_pt
+):
+    mock_ml_lib = p_test_dir / "some/ml/lib"
+    mock_ml_lib.mkdir(parents=True)
+    rai_builder = build.RedisAIBuilder(
+        build_tf=build_tf,
+        build_torch=build_pt,
+        build_onnx=False,
+        libtf_dir=str(mock_ml_lib if build_tf else ""),
+        torch_dir=str(mock_ml_lib if build_pt else ""),
+    )
+    needed_deps = rai_builder._get_deps_to_fetch_for(device)
+    assert len(needed_deps) == 1
+    (dl_pack,) = needed_deps
+    assert isinstance(dl_pack, build._DLPackRepository)
+
+
+def test_rai_builder_raises_if_it_fetches_an_unepected_number_of_ml_deps(
+    monkeypatch, p_test_dir
+):
+    monkeypatch.setattr(
+        build.RedisAIBuilder, "rai_build_path", property(lambda self: p_test_dir)
+    )
+    monkeypatch.setattr(
+        build,
+        "_place_rai_dep_at",
+        lambda target, verbose: lambda dep: target
+        / "whoops_all_ml_deps_extract_to_a_dir_with_this_name",
+    )
+    rai_builder = build.RedisAIBuilder(build_tf=True, build_torch=True, build_onnx=True)
+    with pytest.raises(
+        build.BuildError,
+        match=r"Expected to place \d+ dependencies, but only found \d+",
+    ):
+        rai_builder._fetch_deps_for("cpu")
+
+
+def test_threaded_map():
+    def _some_io_op(x):
+        return x * x
+
+    assert (0, 1, 4, 9, 16) == tuple(build._threaded_map(_some_io_op, range(5)))
+
+
+def test_threaded_map_returns_early_if_nothing_to_map():
+    def _some_long_io_op(_):
+        time.sleep(60)
+
+    start = time.time()
+    build._threaded_map(_some_long_io_op, [])
+    end = time.time()
+    assert end - start < 0.5
```

Comment:
Just to make sure everything is consistent
```suggestion
def test_threaded_map_returns_early_if_nothing_to_map():
    sleep_duration = 60
    def _some_long_io_op(_):
        time.sleep(sleep_duration)

    start = time.time()
    build._threaded_map(_some_long_io_op, [])
    end = time.time()
    assert end - start < sleep_duration
```

Reply:
Good call!